src/common/observablevector.ts

```diff
@@ -0,0 +1,693 @@
+   *
+   * #### Complexity
+   * Linear.
+   *
+   * #### Iterator Validity
+   * No changes.
+   */
+  pushAll(values: IterableOrArrayLike<T>): number;
+
+  /**
+   * Insert a set of items into the vector at the specified index.
+   *
+   * @param index - The index at which to insert the values.
+   *
+   * @param values - The values to insert at the specified index.
+   *
+   * @returns The new length of the vector.
+   *
+   * #### Complexity.
+   * Linear.
+   *
+   * #### Iterator Validity
+   * No changes.
+   *
+   * #### Notes
+   * The `index` will be clamped to the bounds of the vector.
+   *
+   * #### Undefined Behavior.
+   * An `index` which is non-integral.
+   */
+  insertAll(index: number, values: IterableOrArrayLike<T>): number;
+
+  /**
+   * Remove a range of items from the list.
+   *
+   * @param startIndex - The start index of the range to remove (inclusive).
+   *
+   * @param endIndex - The end index of the range to remove (exclusive).
+   *
+   * @returns The new length of the vector.
+   *
+   * #### Complexity
+   * Linear.
+   *
+   * #### Iterator Validity
+   * Iterators pointing to the first removed value and beyond are invalid.
+   *
+   * #### Undefined Behavior
+   * A `startIndex` or `endIndex` which is non-integral.
+   */
+  removeRange(startIndex: number, endIndex: number): number;
+}
+
+
+/**
+ * A concrete implementation of [[IObservableVector]].
+ */
+export
+class ObservableVector<T> extends Vector<T> implements IObservableVector<T> {
+  /**
+   * A signal emitted when the list has changed.
+   */
+  changed: ISignal<ObservableVector<T>, ObservableVector.IChangedArgs<T>>;
+
+  /**
+   * Test whether the list has been disposed.
+   */
+  get isDisposed(): boolean {
+    return this._isDisposed;
+  }
+
+  /**
+   * Dispose of the resources held by the list.
+   */
+  dispose(): void {
+    if (this.isDisposed) {
+      return;
+    }
+    this._isDisposed = true;
+    this.clear();
+    clearSignalData(this);
+  }
+
+  /**
+   * Set the value at the specified index.
+   *
+   * @param index - The positive integer index of interest.
+   *
+   * @param value - The value to set at the specified index.
+   *
+   * #### Complexity
+   * Constant.
+   *
+   * #### Iterator Validity
+   * No changes.
+   *
+   * #### Undefined Behavior
+   * An `index` which is non-integral or out of range.
+   */
+  set(index: number, item: T): void {
+    let oldValues = new Vector([this.at(index)]);
+    super.set(index, item);
+    this.changed.emit({
+      type: 'set',
+      oldIndex: index,
+      newIndex: index,
+      oldValues,
+      newValues: new Vector([item])
+    });
+  }
+
+  /**
+   * Add a value to the back of the vector.
+   *
+   * @param value - The value to add to the back of the vector.
+   *
+   * @returns The new length of the vector.
+   *
+   * #### Complexity
+   * Constant.
+   *
+   * #### Iterator Validity
+   * No changes.
+   */
+  pushBack(value: T): number {
```

Comment:
For this and *all* of the functions that mutate our underlying data structure *and also have a return value*, I'm not sure, but we might want to `requestAnimationFrame` and defer the emission of a signal until after this function returns, because it's a little unclear when a secondary behavior that is a byproduct of this mutation should happen.

Actually, it's possible we may want this even with those functions that mutate without a return value (like `set` and `clear`).

Maybe something like:

```typescript
pushBack(value: T): void {
  let length = super.pushBack(value);
  super.set(index, value);
  this._deferEmission({
    type: 'add',
    oldIndex: -1,
    newIndex: this.length - 1,
    oldValues: this._emptySequence,
    newValues: new Vector([value])
  });
  return length;
}

/* ... */

private _deferEmission(args: ObservableVector.IChangedArgs<T>): void {
  requestAnimationFrame(() => { this.changed.emit(args); });
}
```